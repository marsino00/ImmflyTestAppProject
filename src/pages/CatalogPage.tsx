import React, { useEffect } from 'react';
import { View, ActivityIndicator, Text } from 'react-native';
import { useAppDispatch, useAppSelector } from '../store/hooks';
import { loadProducts } from '../store/slices/catalog.slice';
import CategoryAccordionList from '../components/CategoryAccordionList/CategoryAccordionList';
import CartList from '../components/CartList/CartList';

export default function CatalogPage() {
  const dispatch = useAppDispatch();
  const { products, loading, error } = useAppSelector(s => s.catalog);

  useEffect(() => {
    dispatch(loadProducts());
  }, [dispatch]);

  if (loading) return <ActivityIndicator size="large" />;
  if (error) return <Text>Error: {error}</Text>;

  return (
    <View>
      <CategoryAccordionList products={products} singleOpen={true} />
      <CartList />
    </View>
  );
}
