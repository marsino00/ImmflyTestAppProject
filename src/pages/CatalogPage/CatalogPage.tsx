import React, { useEffect } from 'react';
import { View, ActivityIndicator, Text } from 'react-native';
import { useAppDispatch, useAppSelector } from '../../store/hooks';
import { loadProducts } from '../../store/slices/catalog.slice';
import CategoryAccordionList from '../../components/CategoryAccordionList/CategoryAccordionList';
import CartList from '../../components/CartList/CartList';
import { styles } from './CatalogPage.styles';

export default function CatalogPage() {
  const dispatch = useAppDispatch();
  const { products, loading, error } = useAppSelector(s => s.catalog);

  useEffect(() => {
    dispatch(loadProducts());
  }, [dispatch]);

  if (loading)
    return <ActivityIndicator testID="loading-indicator" size="large" />;
  if (error) return <Text>Error: {error}</Text>;

  return (
    <View style={styles.container}>
      <CategoryAccordionList products={products} singleOpen={true} />
      <CartList />
    </View>
  );
}
