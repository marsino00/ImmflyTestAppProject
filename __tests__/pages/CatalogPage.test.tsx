import React from 'react';
import { render, screen } from '@testing-library/react-native';
import CatalogPage from '../../src/pages/CatalogPage/CatalogPage';

jest.mock(
  '../../src/components/CategoryAccordionList/CategoryAccordionList',
  () => {
    const { View } = require('react-native');
    return function MockCategoryAccordionList() {
      return <View testID="mock-category-accordion-list" />;
    };
  },
);

jest.mock('../../src/components/CartList/CartList', () => {
  const { View } = require('react-native');
  return function MockCartList() {
    return <View testID="mock-cart-list" />;
  };
});

type CatalogSlice = {
  products: any[];
  loading: boolean;
  error: string | null;
};

let mockCatalogState: CatalogSlice = {
  products: [],
  loading: false,
  error: null,
};

const setMockCatalogState = (partial: Partial<CatalogSlice>) => {
  mockCatalogState = { ...mockCatalogState, ...partial };
};

const mockDispatch = jest.fn();

jest.mock('../../src/store/hooks', () => ({
  useAppDispatch: () => mockDispatch,
  useAppSelector: (selector: any) =>
    selector({
      catalog: mockCatalogState,
      cart: { items: {} },
    }),
}));

const loadProductsAction = { type: 'catalog/loadProducts' };
jest.mock('../../src/store/slices/catalog.slice', () => ({
  loadProducts: jest.fn(() => loadProductsAction),
}));

jest.mock('../../src/pages/CatalogPage/CatalogPage.styles', () => ({
  styles: { container: {} },
}));

describe('CatalogPage', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    setMockCatalogState({ products: [], loading: false, error: null });
  });

  it('should dispatch loadProducts on mount', () => {
    render(<CatalogPage />);
    const { loadProducts } = require('../../src/store/slices/catalog.slice');
    expect(loadProducts).toHaveBeenCalledTimes(1);
    expect(mockDispatch).toHaveBeenCalledWith(loadProductsAction);
  });

  it('should show ActivityIndicator when loading is true', () => {
    setMockCatalogState({ loading: true });
    render(<CatalogPage />);
    expect(screen.getByTestId('loading-indicator')).toBeTruthy();
  });

  it('should show error message when error is not null', () => {
    setMockCatalogState({ error: 'Failed to load' });
    render(<CatalogPage />);
    expect(screen.getByText(/Error: Failed to load/i)).toBeTruthy();
  });

  it('should render CategoryAccordionList and CartList when data is loaded without errors', () => {
    setMockCatalogState({
      products: [{ id: 1, name: 'P1', category: 'Drinks', prices: {} }],
      loading: false,
      error: null,
    });
    const { toJSON } = render(<CatalogPage />);
    expect(screen.queryByTestId('loading-indicator')).toBeNull();
    expect(screen.queryByText(/Error:/)).toBeNull();
    expect(screen.getByTestId('mock-category-accordion-list')).toBeTruthy();
    expect(screen.getByTestId('mock-cart-list')).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });
});
