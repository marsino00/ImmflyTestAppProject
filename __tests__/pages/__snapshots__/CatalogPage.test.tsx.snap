// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatalogPage should render CategoryAccordionList and CartList when data is loaded without errors 1`] = `
<View
  style={{}}
>
  <View
    testID="mock-category-accordion-list"
  />
  <View
    testID="mock-cart-list"
  />
</View>
`;
